### PR TITLE
New Integral Invariant functor

### PR DESCRIPTION
**Pull request by dcoeurjo**
# PR Description

Minor PR to retrieve the principal curvature directions and values at the same time.

# Checklist

- [x] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [x] Doxygen documentation of the code completed (classes, methods, types, members...)
- [x] Documentation module page added or updated.
- [x] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [x] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [x] All continuous integration tests pass (Travis & appveyor)

### CODE REVIEW

**Comment by dcoeurjo** [comment]
This one is ready for review.

**Comment by dcoeurjo** [comment]
Can someone review this PR?

**Comment by rolanddenis** [comment]
> Can someone review this PR, plz?

@dcoeurjo I can take a look wednesday :wink:

File: src/DGtal/helpers/ShortcutsGeometry.h [hunk 1]
```diff
@@ -976,6 +981,132 @@ namespace DGtal
...
+        ///   - verbose         [     1]: verbose trace mode 0: silent, 1: verbose.
+        ///   - r-radius        [   3.0]: the constant for kernel radius parameter r in r(h)=r h^alpha (VCM,II,Trivial).
+        ///   - alpha           [  0.33]: the parameter alpha in r(h)=r h^alpha (VCM, II)."
+        ///   - gridstep        [   1.0]: the digitization gridstep (often denoted by h).
+        ///
+        /// @return the vector containing the estimated principal curvatures and directions,
+        ///  in the same order as \a surfels.
+        ///
+        /// @note The function is faster when surfels are in a specific order, as
+        /// given for instance by a depth-first traversal (@see getSurfelRange)
```
Comment (rolanddenis) [inline src/DGtal/helpers/ShortcutsGeometry.h]:
```suggestion
        /// given for instance by a depth-first traversal (see @ref getSurfelRange)
```

File: src/DGtal/helpers/ShortcutsGeometry.h [hunk 2]
```diff
@@ -976,6 +981,132 @@ namespace DGtal
           return mc_estimations;
         }
 
+      
+        /// Given a digital shape \a bimage, a sequence of \a surfels,
```
Comment (rolanddenis) [inline src/DGtal/helpers/ShortcutsGeometry.h]: Please remove two spaces of indentation to be aligned with previous functions

File: src/DGtal/helpers/ShortcutsGeometry.h [hunk 3]
```diff
@@ -976,6 +981,132 @@ namespace DGtal
...
+          return getIIPrincipalCurvaturesAndDirections( *bimage, K, surfels, params );
+        }
+
+      /// Given a digital shape \a bimage, a sequence of \a surfels,
+      /// and some parameters \a vm, returns the principal curvatures and
+      /// directions using an Integral
+      /// Invariant (II) estimation at the specified surfels, in the
+      /// same order.
+      ///
+        /// @param[in] dshape the digitized implicit shape, which is an
```
Comment (rolanddenis) [inline src/DGtal/helpers/ShortcutsGeometry.h]: Change of indentation width detected :octocat:

File: src/DGtal/helpers/ShortcutsGeometry.h [hunk 4]
```diff
@@ -976,6 +981,132 @@ namespace DGtal
...
+        /// @param[in] surfels the sequence of surfels at which we compute the normals
+        /// @param[in] params the parameters:
+        ///   - verbose         [     1]: verbose trace mode 0: silent, 1: verbose.
+        ///   - r-radius        [   3.0]: the constant for kernel radius parameter r in r(h)=r h^alpha (VCM,II,Trivial).
+        ///   - alpha           [  0.33]: the parameter alpha in r(h)=r h^alpha (VCM, II)."
+        ///   - gridstep        [   1.0]: the digitization gridstep (often denoted by h).
+        ///   - minAABB         [ -10.0]: the min value of the AABB bounding box (domain)
+        ///   - maxAABB         [  10.0]: the max value of the AABB bounding box (domain)
+        ///   - offset          [   5.0]: the digital dilation of the digital space,
+        ///                       useful when you process shapes and that you add noise.
```
Comment (rolanddenis) [inline src/DGtal/helpers/ShortcutsGeometry.h]: Is this sentence correct?

File: tests/geometry/surfaces/testIntegralInvariantCovarianceEstimator.cpp [hunk 5]
```diff
@@ -203,22 +209,40 @@ bool testPrincipalCurvatures3d( double h )
...
+  
+  
+  trace.beginBlock( "Checking CurvaturesAndDirections functor");
+  Value val = curvatureEstimator.eval( surf.begin() );
+  ValueTensor valTensor = curvatureEstimatorTensor.eval( surf.begin() );
+  bool ok = (val.first ==  std::get<0>(valTensor));
+  ok &= (val.second ==  std::get<1>(valTensor));
+  if (!ok)
+    trace.error()<< "Error comparing principal curvatures between two different functors."<<std::endl;
+
```
Comment (rolanddenis) [inline tests/geometry/surfaces/testIntegralInvariantCovarianceEstimator.cpp]: Shouldn't we test the estimation values also?

File: src/DGtal/geometry/surfaces/estimation/IIGeometricFunctors.h [hunk 6]
```diff
@@ -435,6 +437,92 @@ namespace DGtal {
...
+        * Euclidean radius. Not used for this estimator.
+         */
+        void init( Component h , Component  r )
+        {
+          double r3 = r * r * r;
+          double r6 = r3 * r3;
+          d6_PIr6 = 6.0 / ( M_PI * r6 );
+          d8_5r = 8.0 / ( 5.0 * r );
+          double h2 = h * h;
+          dh5 = h2 * h2 * h;
```
[the remark this answers is not in the document]
Comment (dcoeurjo) [reply src/DGtal/geometry/surfaces/estimation/IIGeometricFunctors.h]: Yep, this formula only makes sense for surfaces in R^3 (as for the notion of k1, k2 principal directions).

File: ChangeLog.md [hunk 7]
```diff
@@ -18,11 +18,15 @@
   - DGtal can be compiled and used as a project (git) submodule (David
   Coeurjolly [#1444](https://github.com/DGtal-team/DGtal/pull/1444))
   - Add .gitattributes file for github to recognize ih files as c++
-    (Pablo Hernandez-Cerdan [#1457](https://github.com/DGtal-team/DGtal/pull/1457))
-  
+    (Pablo Hernandez-Cerdan [#1460](https://github.com/DGtal-team/DGtal/pull/1460))
```
[the remark this answers is not in the document]
Comment (dcoeurjo) [reply ChangeLog.md]: oops thx

Comment (dcoeurjo) [inline ChangeLog.md]:
```suggestion
    (Pablo Hernandez-Cerdan [#1457](https://github.com/DGtal-team/DGtal/pull/1457))
```

File: src/DGtal/helpers/ShortcutsGeometry.h (hunk 2 above)
Comment (dcoeurjo) [reply src/DGtal/helpers/ShortcutsGeometry.h]: Thx

File: src/DGtal/helpers/ShortcutsGeometry.h (hunk 3 above)
Comment (dcoeurjo) [reply src/DGtal/helpers/ShortcutsGeometry.h]: 👍

File: tests/geometry/surfaces/testIntegralInvariantCovarianceEstimator.cpp (hunk 5 above)
Comment (dcoeurjo) [reply tests/geometry/surfaces/testIntegralInvariantCovarianceEstimator.cpp]: No need I think. The first functor is already tested. We just make sure that the approximations coincide

File: src/DGtal/helpers/Shortcuts.h [hunk 8]
```diff
@@ -189,6 +190,9 @@ namespace DGtal
       typedef GradientColorMap<Scalar>                            ColorMap;
       typedef TickedColorMap<Scalar,ColorMap>                     ZeroTickedColorMap;
     
+      typedef std::tuple< double,double, RealVector,RealVector>   CurvatureTensorQuantity;
+      typedef std::vector< CurvatureTensorQuantity >              CurvatureTensorQuantities;
```
[the remark this answers is not in the document]
Comment (dcoeurjo) [reply src/DGtal/helpers/Shortcuts.h]: I agree.. thx

File: src/DGtal/geometry/surfaces/estimation/IIGeometricFunctors.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (rolanddenis) [reply src/DGtal/geometry/surfaces/estimation/IIGeometricFunctors.h]: Ok, so we should reflect this in the documentation and space dimension test

File: src/DGtal/geometry/surfaces/estimation/IIGeometricFunctors.h [hunk 9]
```diff
@@ -435,6 +437,92 @@ namespace DGtal {
...
+        typedef typename Space::RealVector RealVector;
+        typedef typename RealVector::Component Component;
+        typedef TMatrix Matrix;
+        typedef Matrix Argument;
+        typedef std::tuple<double, double, RealVector, RealVector> Quantity;
+        typedef Quantity Value;
+
+        BOOST_CONCEPT_ASSERT(( concepts::CMatrix<Matrix> ));
+        BOOST_CONCEPT_ASSERT(( concepts::CSpace<TSpace> ));
+        BOOST_STATIC_ASSERT(( Space::dimension >= 3 ));
```
Comment (rolanddenis) [inline src/DGtal/geometry/surfaces/estimation/IIGeometricFunctors.h]: Here:
```suggestion
        BOOST_STATIC_ASSERT(( Space::dimension == 3 ));
```

File: src/DGtal/geometry/surfaces/estimation/IIGeometricFunctors.h [hunk 10]
```diff
@@ -435,6 +437,92 @@ namespace DGtal {
...
 
+  
+      /////////////////////////////////////////////////////////////////////////////
+      // template class IIPrincipalCurvaturesAndDirectionsFunctor
+      /**
+      * Description of template class
+      * 'IIPrincipalCurvaturesAndDirectionsFunctor' <p> \brief Aim: A functor
+      * Matrix -> std::pair<RealVector,RealVector> that returns the first and
+      * the second principal curvature directions by diagonalizing the given
+      * covariance matrix. This functor is valid starting from 3D space.
```
Comment (rolanddenis) [inline src/DGtal/geometry/surfaces/estimation/IIGeometricFunctors.h]: And here:
```suggestion
      * covariance matrix. This functor is valid only for 3D space.
```

File: src/DGtal/geometry/surfaces/estimation/IIGeometricFunctors.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (dcoeurjo) [reply src/DGtal/geometry/surfaces/estimation/IIGeometricFunctors.h]: Agree.. let met check the existing functors..

File: src/DGtal/geometry/surfaces/estimation/IIGeometricFunctors.h (hunk 9 above)
Comment (dcoeurjo) [reply src/DGtal/geometry/surfaces/estimation/IIGeometricFunctors.h]: thx!

File: src/DGtal/helpers/Shortcuts.h [hunk 11]
```diff
@@ -189,6 +190,9 @@ namespace DGtal
       typedef GradientColorMap<Scalar>                            ColorMap;
       typedef TickedColorMap<Scalar,ColorMap>                     ZeroTickedColorMap;
     
+      typedef std::tuple< double,double, RealVector,RealVector>   CurvatureTensorQuantity;
```
[the remark this answers is not in the document]
Comment (dcoeurjo) [reply src/DGtal/helpers/Shortcuts.h]: cleaned up

File: src/DGtal/helpers/ShortcutsGeometry.h (hunk 4 above)
Comment (dcoeurjo) [reply src/DGtal/helpers/ShortcutsGeometry.h]: I have rephrased

**Comment by dcoeurjo** [comment]
It should be ok. Thx @rolanddenis for the review ;)

File: tests/geometry/surfaces/testIntegralInvariantCovarianceEstimator.cpp (hunk 5 above)
Comment (dcoeurjo) [reply tests/geometry/surfaces/testIntegralInvariantCovarianceEstimator.cpp]: I've added another test to make sure that the Shortcuts are fine.

File: src/DGtal/helpers/ShortcutsGeometry.h (hunk 4 above)
Comment (rolanddenis) [reply src/DGtal/helpers/ShortcutsGeometry.h]: :+1:

File: tests/geometry/surfaces/testIntegralInvariantCovarianceEstimator.cpp (hunk 5 above)
Comment (rolanddenis) [reply tests/geometry/surfaces/testIntegralInvariantCovarianceEstimator.cpp]: Ok I see now, thx!

**Comment by dcoeurjo** [comment]
thx @rolanddenis for the comments. Can you approve this PR? Are there still pending comments?

**Comment by rolanddenis** [comment]
Approved!

@dcoeurjo did you see the missing jobs (DGtalTools & doc) and the error in configuration (conflict between `matrix` and `jobs`): https://travis-ci.org/github/DGtal-team/DGtal/builds/662684925/config?utm_medium=notification&utm_source=github_status ?

**Comment by dcoeurjo** [comment]
wut ?? travis odd again ?.. damit
thanks a lot for your final commits;)

**Comment by dcoeurjo** [comment]
first half commits in this PR are fine.. and then travis goes wrong...
Merging this one, we'll see travis issue later
thx again

**Pull request merged**